functions/src/pdfBuilder/generateDonationCertificatePDF.ts

```diff
@@ -0,0 +1,156 @@
+		//return await createDefaultDonationCertificate(fileRef, data, translations);
+	}
+
+}
+
+export const createDonationCertificateCH = async (fileRef: any, data: any, translations: Map<string, string>, firestore: any, storage: any) => {
+
+	let pdfData: ArrayBuffer = new ArrayBuffer(8);
+	let buffers: any = [];
+
+	const PDFDocument = require('pdfkit');
+	let docPdf = new PDFDocument();
+
+	//Address for contributor
+	
+	docPdf.fontSize(12).text(
+		`${data.currentDate}`,
+		100, 80);
+	docPdf.moveDown();
+	docPdf.fontSize(12).text(`Social Income\nZwierstrasse 103\n8004 Zürich\nSwitzerland`);
+
+	//Address of social income
+	docPdf.moveDown();
+	docPdf.fontSize(12).text(
+		`${data.personal.name} ${data.personal.lastname}\n${data.address.street}\n${data.address.zip} ${data.address.city}\n${data.address.country}`, {
+	width: 410,
+	align: 'right'
+	}
+	);
+
+	docPdf.moveDown();
+	docPdf.moveDown();
+	docPdf.fontSize(14).text(translations.get('donation-certificate.title') + ' ');
+	docPdf.moveDown();
+	docPdf.fontSize(12).text(translations.get('donation-certificate.salutation') + `${data.personal.name} ${data.personal.lastname}`);
+	docPdf.moveDown();
+	docPdf.fontSize(12).text(
+		translations.get('donation-certificate.confirmation-1')
+		+ data.year
+		+ translations.get('donation-certificate.confirmation-2')
+	);
+
+	if (!(data.financials.total_chf !== '0')) {
+		docPdf.fontSize(12).text('Spenden CHF: ' + data.financials.total_chf);
+	}
+	if (!(data.financials.total_eur !== '0')) {
+		docPdf.fontSize(12).text('Spenden EUR: ' + data.financials.total_chf);
+	}	
+	if (!(data.financials.total_chf !== '0')) {
+		docPdf.fontSize(12).text('Spenden USD: ' + data.financials.total_chf);
+	}
+
+	docPdf.moveDown();
+	docPdf.fontSize(12).text(translations.get('donation-certificate.information-1'));
+	docPdf.moveDown();
+	docPdf.fontSize(12).text(translations.get('donation-certificate.thanks'));
+	docPdf.moveDown();
+	docPdf.fontSize(12).text(translations.get('donation-certificate.greetings'));
+	docPdf.moveDown();
+	docPdf.moveDown();
+	docPdf.moveDown();
+	docPdf.moveDown();
+	docPdf.fontSize(12).text(translations.get('donation-certificate.sender'));
+	docPdf.moveDown();
+	docPdf.moveDown();
+	docPdf.addPage();
+	docPdf.fontSize(12).text(translations.get('donation-certificate.single-donations'));
+	docPdf.moveDown();
+	const lineItems = data.financials.lineItems;
+	for (let i = 0; i < lineItems.length; ++i) {
+		const elements = lineItems[i].split(';');
+		docPdf.fontSize(12).text(elements[0] + ': ' + elements[1] + " " + elements[2].toUpperCase());
+	}
+
+	docPdf.on('data', buffers.push.bind(buffers));
+	docPdf.on('end', () => {
+		pdfData = Buffer.concat(buffers);
+		uploadFileAndCreateDocumentEntry(fileRef, pdfData, data, firestore, storage);
+	});
+
+	docPdf.end();	
+}
+
+
+export const addDocumentToUser = async(data: any, fileRef: any, firestore: any, storage: any) => {
+
+	const downloadUrl = '[File Download](' +  await getDownloadURL(fileRef) + ')';
```

Comment:
Have you already put some thoughts into how we ensure that one can only download the certificates of oneself? 
Not that one can get donation information of others by reverse engineer the urls.

Reply:
Could this be also solved with the approach we discussed around the survey links? -> Using custom tokens?

[Create custom tokens using the Firebase Admin SDK](https://firebase.google.com/docs/auth/admin/create-custom-tokens#create_custom_tokens_using_the_firebase_admin_sdk)

> We can create the token when generating the survey links. The token contains the recipient's uuid. The react app can then call the auth endpoint firebase.auth().signInWithCustomToken(token). Once authenticated it can query the survey collection. In the request, the token is passed along and one can read the recipient's uuid in the security rules. I would go for this approach since jwt token authentication is a standard and widely used.

Reply:
The donation certificates use the ID of the user in the url and I attached an additional random string at the end to make sure reverse engineering is not possible.